File: bindings/python/api/rest_api.py
"""
REST API client for Satox.
"""

import json
import time
from typing import Dict, Any, Optional
import requests
from requests.exceptions import RequestException

class APIError(Exception):
    """Base class for API errors."""
    def __init__(self, message: str, status_code: Optional[int] = None):
        self.message = message
        self.status_code = status_code
        super().__init__(f"API Error: {message} (Status: {status_code})")

class APIResponse:
    """Represents an API response."""
    def __init__(self, response: requests.Response):
        self.status_code = response.status_code
        self.headers = response.headers
        self.data = response.json() if response.content else None

class RESTAPIClient:
    """REST API client for Satox."""
    
    def __init__(self, base_url: str, api_key: str):
        self.base_url = base_url.rstrip('/')
        self.api_key = api_key
        self.session = requests.Session()
        self.session.headers.update(self._get_headers())

    def _get_headers(self) -> Dict[str, str]:
        """Get request headers."""
        return {
            'Authorization': f'Bearer {self.api_key}',
            'Content-Type': 'application/json',
            'Accept': 'application/json'
        }

    def _handle_response(self, response: requests.Response) -> APIResponse:
        """Handle API response."""
        if response.status_code >= 400:
            try:
                error_data = response.json()
                error_message = error_data.get('error', 'Unknown error')
            except ValueError:
                error_message = response.text or 'Unknown error'
            raise APIError(error_message, response.status_code)
        return APIResponse(response)

    def get(self, endpoint: str, params: Optional[Dict[str, Any]] = None) -> APIResponse:
        """Make a GET request."""
        url = f"{self.base_url}/{endpoint.lstrip('/')}"
        try:
            response = self.session.get(url, params=params)
            return self._handle_response(response)
        except RequestException as e:
            raise APIError(str(e), 503)

    def post(self, endpoint: str, data: Dict[str, Any]) -> APIResponse:
        """Make a POST request."""
        url = f"{self.base_url}/{endpoint.lstrip('/')}"
        try:
            response = self.session.post(url, json=data)
            return self._handle_response(response)
        except RequestException as e:
            raise APIError(str(e), 503)

    def put(self, endpoint: str, data: Dict[str, Any]) -> APIResponse:
        """Make a PUT request."""
        url = f"{self.base_url}/{endpoint.lstrip('/')}"
        try:
            response = self.session.put(url, json=data)
            return self._handle_response(response)
        except RequestException as e:
            raise APIError(str(e), 503)

    def delete(self, endpoint: str) -> APIResponse:
        """Make a DELETE request."""
        url = f"{self.base_url}/{endpoint.lstrip('/')}"
        try:
            response = self.session.delete(url)
            return self._handle_response(response)
        except RequestException as e:
            raise APIError(str(e), 503)

    def close(self):
        """Close the session."""
        self.session.close() 

File: bindings/go/satox/api/rest_api.go
package api

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"sync"
	"time"
)

type RESTAPI struct {
	mu sync.Mutex
}

func NewRESTAPI() *RESTAPI {
	return &RESTAPI{}
}

func (r *RESTAPI) Initialize() error {
	return nil
}

func (r *RESTAPI) Shutdown() error {
	return nil
}

func (r *RESTAPI) StartServer() error {
	r.mu.Lock()
	defer r.mu.Unlock()
	// Implement server start logic here
	return nil
}

func (r *RESTAPI) StopServer() error {
	r.mu.Lock()
	defer r.mu.Unlock()
	// Implement server stop logic here
	return nil
}

func (r *RESTAPI) HandleRequest(endpoint string, data []byte) ([]byte, error) {
	r.mu.Lock()
	defer r.mu.Unlock()
	// Implement request handling logic here
	return nil, errors.New("Request handling failed")
}

// RestAPIClient represents a REST API client
type RestAPIClient struct {
	baseURL    string
	httpClient *http.Client
	apiKey     string
}

// NewRestAPIClient creates a new REST API client
func NewRestAPIClient(baseURL string, apiKey string) *RestAPIClient {
	return &RestAPIClient{
		baseURL: baseURL,
		httpClient: &http.Client{
			Timeout: time.Second * 30,
		},
		apiKey: apiKey,
	}
}

// Request represents an API request
type Request struct {
	Method  string
	Path    string
	Body    interface{}
	Headers map[string]string
}

// Response represents an API response
type Response struct {
	StatusCode int
	Body       []byte
	Headers    http.Header
}

// Do sends an API request and returns the response
func (c *RestAPIClient) Do(ctx context.Context, req Request) (*Response, error) {
	var bodyReader io.Reader
	if req.Body != nil {
		jsonBody, err := json.Marshal(req.Body)
		if err != nil {
			return nil, fmt.Errorf("failed to marshal request body: %w", err)
		}
		bodyReader = bytes.NewReader(jsonBody)
	}

	httpReq, err := http.NewRequestWithContext(ctx, req.Method, c.baseURL+req.Path, bodyReader)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	// Set default headers
	httpReq.Header.Set("Content-Type", "application/json")
	httpReq.Header.Set("X-API-Key", c.apiKey)

	// Set custom headers
	for key, value := range req.Headers {
		httpReq.Header.Set(key, value)
	}

	resp, err := c.httpClient.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response body: %w", err)
	}

	return &Response{
		StatusCode: resp.StatusCode,
		Body:       body,
		Headers:    resp.Header,
	}, nil
}

// Get sends a GET request
func (c *RestAPIClient) Get(ctx context.Context, path string, headers map[string]string) (*Response, error) {
	return c.Do(ctx, Request{
		Method:  http.MethodGet,
		Path:    path,
		Headers: headers,
	})
}

// Post sends a POST request
func (c *RestAPIClient) Post(ctx context.Context, path string, body interface{}, headers map[string]string) (*Response, error) {
	return c.Do(ctx, Request{
		Method:  http.MethodPost,
		Path:    path,
		Body:    body,
		Headers: headers,
	})
}

// Put sends a PUT request
func (c *RestAPIClient) Put(ctx context.Context, path string, body interface{}, headers map[string]string) (*Response, error) {
	return c.Do(ctx, Request{
		Method:  http.MethodPut,
		Path:    path,
		Body:    body,
		Headers: headers,
	})
}

// Delete sends a DELETE request
func (c *RestAPIClient) Delete(ctx context.Context, path string, headers map[string]string) (*Response, error) {
	return c.Do(ctx, Request{
		Method:  http.MethodDelete,
		Path:    path,
		Headers: headers,
	})
}

// Patch sends a PATCH request
func (c *RestAPIClient) Patch(ctx context.Context, path string, body interface{}, headers map[string]string) (*Response, error) {
	return c.Do(ctx, Request{
		Method:  http.MethodPatch,
		Path:    path,
		Body:    body,
		Headers: headers,
	})
}

// SetTimeout sets the client timeout
func (c *RestAPIClient) SetTimeout(timeout time.Duration) {
	c.httpClient.Timeout = timeout
}

// SetAPIKey sets the API key
func (c *RestAPIClient) SetAPIKey(apiKey string) {
	c.apiKey = apiKey
}

// SetBaseURL sets the base URL
func (c *RestAPIClient) SetBaseURL(baseURL string) {
	c.baseURL = baseURL
}
